\documentclass{article}

<<style, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\begin{document}

\title{massay}
\author{Bioconductor MultiAssay SIG}
\maketitle

\tableofcontents

\section{Introduction}

An increasingly common use case involves a set of samples or patients who provide measurements on
multiple data types, such as gene expression, genotype, miRNA abundance.  It will frequently be the
case that not all samples will contribute to all assays, so some sparsity in the set of samples
$\times$ assays is expected.

<<>>=
library(massay)
@ 

\section{Basic demonstrative resources}

\subsubsection*{Creating a MultiAssayExperiment}

This should be done using constructors:

<<createeHub>>=
rdas <- list.files(system.file("extdata/tcga_ov", package="biocMultiAssay"), pattern = "\\.rda$", full = TRUE)
basename(rdas)
pheno <- grep("phenodata\\.rda$", rdas, value = TRUE)
rdas <- rdas[ - grep("phenodata\\.rda$", rdas)]
tags <- sub("\\.rda", "", basename(rdas))
eList <- lapply(seq(along = rdas), function(ii) {
    new("SerializedExperiment", serType = "RData", assayPath = rdas[ii], tag = tags[ii], sampleDataPath = rdas[ii])
})
mAexp <- new("MultiAssayExperiment", experiments = eList, sampleData = as(get(load(pheno)), "DataFrame"))
@

\subsubsection*{Loading a MultiAssayExperiment}

<<view>>=
mAexp
@

<<loadHub>>=
mAexp <- loadExperiments(mAexp)
mAexp
@ 

(Want to support loading of only parts of the hub).

\subsubsection*{Subsetting a MultiAssayExperiment}

<<subsetBySample>>=
subsetBySample(mAexp, 1:10)
@ 

<<selectExperiments>>=
selectExperiments(mAexp, 1:3)
## Should implement tag subsetting
@ 

<<>>=
getExperiment(mAexp, 1)
getExperiment(mAexp, "TCGA.RNASeqV2_eset")
@ 


\section{Session information}

<<sessionInfo,results=tex,echo=FALSE>>=
toLatex(sessionInfo())
@

\end{document}
